### PR TITLE
Add tutorial to build and containerise go application

### PR DESCRIPTION
**Pull request by kameshsampath**
# Harness Developer Pull Request
Thanks for helping us make the Developer Hub better. The PR will be looked at
by the maintainers. 

## What Type of PR is This?

- [ ] Issue
- [x] Feature
- [ ] Maintenance/Chore

Adding CI tutorial on how to build and containerise go application using CI pipelines.

## House Keeping
Some items to keep track of. Screen shots of changes are optional but would help the maintainers review quicker. 

- [x] Tested Locally
- [x] *Optional* Screen Shoot. 
-

image

image

image

### CODE REVIEW

**Review by ravilach — CHANGES_REQUESTED** [review]
Looks good. Could we drop "CI" from the Page Title and Sidebar Nav?  e.g just " Build Go App". Cheers!
image

**Comment by kameshsampath** [comment]
> Looks good. Could we drop "CI" from the Page Title and Sidebar Nav? e.g just " Build Go App". Cheers! CC @kameshsampath image

done!

**Comment by kameshsampath** [comment]
> One more comment @kameshsampath from my side, can you add your Tile one more spot, in the "All Tutorials" section under CI -> https://github.com/harness/developer-hub/blob/main/src/components/LandingPage/AllTutorials.tsx. Cheers!

done

**Comment by ravilach** [comment]
Thanks @kameshsampath! With a review from @KimberlyFields we should be good!

**Comment by ravilach** [comment]
@kameshsampath looking at the build, on index.tsx, can you make one change? Or can bring your PR up to date with the repo e.g re-base? Not sure why this change didn't proprgate. 

image


```

Exhaustive list of all broken links found:

- On source page path = /:

   -> linking to /tutorials/get-started

```

File:
https://github.com/harness/developer-hub/blob/main/src/pages/index.tsx

Change:

```
     <div className={styles.btnContainer}>
        <Link href="/tutorials">
          <button className="button button--primary">All Tutorials</button>

```

Cheers,

-Ravi

CC: @KimberlyFields @wei-harness

**Comment by kameshsampath** [comment]
@ravilach did a rebase and pushed it back. Hope it is good now

**Comment by ravilach** [comment]
Thanks @kameshsampath, all up to date, PR build passes. Cheers!

**Comment by ravilach** [comment]
Heads up any suggestions please bulk-commit/accept. Each commit to the PR is a build :-)

File: tutorials/build-code/ci-tutorial-container-signing.md [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+
+:::important
+
+Make sure the `cosign.key`, `cosign.pub` are excluded from your SCM i.e. add them to `.gitignore`
+
+:::
+
+### Create Signing Secrets
+
+Navigate to __Project Setup__ --> __Secrets__,
```
Comment (KimberlyFields) [inline tutorials/build-code/ci-tutorial-container-signing.md:68]:
```suggestion
1. Navigate to __Project Setup__, and then select  __Secrets__.
```

File: tutorials/build-code/ci-tutorial-container-signing.md [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+
+![Project Secrets](static/ci-tutorial-container-signing/fruits_api_secrets_list_0.png)
+
+Click __+ New Secret__ and select __Text__,
```
Comment (KimberlyFields) [inline tutorials/build-code/ci-tutorial-container-signing.md:72]:
```suggestion
2. Select __+ New Secret__ , and then select __Text__.
```

File: tutorials/build-code/ci-tutorial-container-signing.md [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+
+![New Text Secret](static/ci-tutorial-container-signing/new_text_secret_cosign_password.png)
+
+Set the __Secret Name__ to be `cosign_password` and fill your cosign private key password i.e. `$COSIGN_PASSSWORD` as __Secret Value__,
```
Comment (KimberlyFields) [inline tutorials/build-code/ci-tutorial-container-signing.md:76]:
```suggestion
3. Set the __Secret Name__ to be `cosign_password`, and then fill your cosign private key password `$COSIGN_PASSSWORD` as __Secret Value__.
```

File: tutorials/build-code/ci-tutorial-container-signing.md [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+
+![Cosign Password](static/ci-tutorial-container-signing/secret_cosign_password.png)
+
+Click __+ New Secret__ and select __File__,
```
Comment (KimberlyFields) [inline tutorials/build-code/ci-tutorial-container-signing.md:80]:
```suggestion
4. Select __+ New Secret__, and then select __File__.
```

File: tutorials/build-code/ci-tutorial-container-signing.md [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+
+![New File Secret](static/ci-tutorial-container-signing/new_file_secret.png)
+
+Fill the details for the secret `cosign_private_key` as shown,
```
Comment (KimberlyFields) [inline tutorials/build-code/ci-tutorial-container-signing.md:84]:
```suggestion
5. Fill the details for the secret `cosign_private_key` as shown.
```

File: tutorials/build-code/ci-tutorial-container-signing.md [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+
+![Cosign Private Key Secret](static/ci-tutorial-container-signing/file_secret_cosign_private_key.png)
+
+For the __Select File__ browse and pick the `cosign.key` from the `$TUTORIAL_HOME`.
```
Comment (KimberlyFields) [inline tutorials/build-code/ci-tutorial-container-signing.md:88]:
```suggestion
6. For the __Select File__, browse and pick the `cosign.key` from the `$TUTORIAL_HOME`.
```

File: tutorials/build-code/ci-tutorial-container-signing.md [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+
+Repeat the same process of creating another `__File__` secret to store the cosign public key, this time using the file `cosign.pub` from the `$TUTORIAL_HOME`.
```
Comment (KimberlyFields) [inline tutorials/build-code/ci-tutorial-container-signing.md:90]:
```suggestion
7. Repeat the same process of creating another `__File__` secret to store the cosign public key, this time using the file `cosign.pub` from the `$TUTORIAL_HOME`.
```

File: tutorials/build-code/ci-tutorial-container-signing.md [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+
+Fill the details for the secret `cosign_public_key` as shown,
```
Comment (KimberlyFields) [inline tutorials/build-code/ci-tutorial-container-signing.md:92]:
```suggestion
8. Fill the details for the secret `cosign_public_key` as shown.
```

File: tutorials/build-code/ci-tutorial-container-signing.md [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+
+![Cosign Public Key Secret](static/ci-tutorial-container-signing/file_secret_cosign_public_key.png)
+
+With this we now have four secrets on our project,
```
Comment (KimberlyFields) [inline tutorials/build-code/ci-tutorial-container-signing.md:96]:
```suggestion
With this we now have four secrets in our project:
```

File: tutorials/build-code/ci-tutorial-container-signing.md [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+
+![Project Secrets](static/ci-tutorial-container-signing/fruits_api_secrets_list_2.png)
+
+## Update Pipeline to Sign the Container Image
```
Comment (KimberlyFields) [inline tutorials/build-code/ci-tutorial-container-signing.md:100]:
```suggestion
## Update the pipeline to sign the container Image
```

File: tutorials/build-code/ci-tutorial-container-signing.md [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+
+We now have all required resources __cosign_password__, __cosign_private_key__ and __cosign_public_key__ to sign the container image using _cosign_.
```
Comment (KimberlyFields) [inline tutorials/build-code/ci-tutorial-container-signing.md:102]:
```suggestion
We now have all of the required resources: __cosign_password__, __cosign_private_key__ and __cosign_public_key__ to sign the container image using _cosign_.
```

File: tutorials/build-code/ci-tutorial-container-signing.md [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+
+Let us now update the existing __build and push__ step of our `Build Go` pipeline to sign the `fruits-api` container image.
```
Comment (KimberlyFields) [inline tutorials/build-code/ci-tutorial-container-signing.md:104]:
```suggestion
Now, update the existing __build and push__ step of our `Build Go` pipeline to sign the `fruits-api` container image.
```

File: tutorials/build-code/ci-tutorial-container-signing.md [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+
+Navigate to the __Projects__ --> __Pipelines__,
```
Comment (KimberlyFields) [inline tutorials/build-code/ci-tutorial-container-signing.md:106]:
```suggestion
1. Navigate to __Projects__, and then  __Pipelines__.
```

File: tutorials/build-code/ci-tutorial-container-signing.md [hunk 14]
```diff
@@ -0,0 +1,215 @@
...
+
+![Pipelines List](static/ci-tutorial-go-containers/project_pipelines.png)
+
+And click __Build Go__ pipeline to open the pipeline editor,
```
Comment (KimberlyFields) [inline tutorials/build-code/ci-tutorial-container-signing.md:110]:
```suggestion
2. Select __Build Go__ pipeline to open the pipeline editor.
```

File: tutorials/build-code/ci-tutorial-container-signing.md [hunk 15]
```diff
@@ -0,0 +1,215 @@
...
+
+![Build Go Pipeline](static/ci-tutorial-go-containers/select_build_go_pipeline_2.png)
+
+Click the step __build and push__ and update the __Command__ to be as shown,
```
Comment (KimberlyFields) [inline tutorials/build-code/ci-tutorial-container-signing.md:114]:
```suggestion
3. Select the step __build and push__ and update the __Command__ to be as shown:
```

File: tutorials/build-code/ci-tutorial-container-signing.md [hunk 16]
```diff
@@ -0,0 +1,215 @@
...
+echo -n "$DOCKER_HUB_PASSWORD" | ko auth login docker.io -u "$DOCKER_HUB_USERNAME" --password-stdin
+IMAGE_REF=$(ko build --bare --platform linux/amd64 --platform linux/arm64 .)
+cosign sign --key env://COSIGN_PRIVATE_KEY $IMAGE_REF
+```
+
+![Build and Push Step](static/ci-tutorial-container-signing/go_pipeline_step_build_push.png)
+
+We also need to configure few environment variables that are required by `ko` and `cosign` to build and push the signed image to `fruits-api` container repository.
+
+Update the __Environment Variables__ section with following values,
```
Comment (KimberlyFields) [inline tutorials/build-code/ci-tutorial-container-signing.md:128]:
```suggestion
4. Update the __Environment Variables__ section with following values:
```

File: tutorials/build-code/ci-tutorial-container-signing.md [hunk 17]
```diff
@@ -0,0 +1,215 @@
...
+DOCKER_HUB_USERNAME: $DOCKER_HUB_USERNAME
+DOCKER_HUB_PASSWORD: <+secrets.getValue("docker_hub_password")>
+KO_DOCKER_REPO: docker.io/$DOCKER_HUB_USERNAME/fruits-api
+COSIGN_PRIVATE_KEY: <+secrets.getValue("cosign_private_key")>
+COSIGN_PASSWORD: <+secrets.getValue("cosign_password")>
+```
+
+![Build and Push Env](static/ci-tutorial-container-signing/go_pipeline_step_build_push_env_vars.png)
+
+The step also exposes an [output variable](https://docs.harness.io/article/4dot4qo5l4-download-and-copy-artifacts-using-the-command-step#output_variables) called `IMAGE_REF` which could be used by other steps e.g. to verify the signature by using pipeline expression `<+steps.build_and_push.output.outputVariables.IMAGE_REF>`.
```
Comment (KimberlyFields) [inline tutorials/build-code/ci-tutorial-container-signing.md:140]:
```suggestion
The step also exposes an [output variable](https://developer.harness.io/docs/continuous-delivery/cd-execution/cd-general-steps/download-and-copy-artifacts-using-the-command-step) called `IMAGE_REF` which could be used by other steps, for example to verify the signature by using pipeline expression `<+steps.build_and_push.output.outputVariables.IMAGE_REF>`.
```

File: tutorials/build-code/ci-tutorial-container-signing.md [hunk 18]
```diff
@@ -0,0 +1,215 @@
...
+
+:::note Points to Ponder
```
Comment (KimberlyFields) [inline tutorials/build-code/ci-tutorial-container-signing.md:142]:
```suggestion
:::note 
```

File: tutorials/build-code/ci-tutorial-container-signing.md [hunk 19]
```diff
@@ -0,0 +1,215 @@
...
+
+- As marked ensure the `DOCKER_HUB_PASSWORD`,`COSIGN_PRIVATE_KEY` and `COSIGN_PASSWORD` are of type __Expression__
```
Comment (KimberlyFields) [inline tutorials/build-code/ci-tutorial-container-signing.md:144]:
```suggestion
- As marked, ensure the `DOCKER_HUB_PASSWORD`,`COSIGN_PRIVATE_KEY` and `COSIGN_PASSWORD` are of the __Expression__ type.
```

File: tutorials/build-code/ci-tutorial-container-signing.md [hunk 20]
```diff
@@ -0,0 +1,215 @@
...
+- `secrets.getValue` is an expression that allows to get the value from the secret `docker_hub_password`, that was created earlier in the tutorial. Check the [docs](https://developer.harness.io/docs/platform/security/add-use-text-secrets/#step-3-reference-the-encrypted-text-by-identifier) for more info
```
Comment (KimberlyFields) [inline tutorials/build-code/ci-tutorial-container-signing.md:145]:
```suggestion
- `secrets.getValue` is an expression that allows you to get the value from the secret `docker_hub_password` that was created earlier in the tutorial. Check the [docs](https://developer.harness.io/docs/platform/security/add-use-text-secrets/#step-3-reference-the-encrypted-text-by-identifier) for more information. 
```

File: tutorials/build-code/ci-tutorial-container-signing.md [hunk 21]
```diff
@@ -0,0 +1,215 @@
...
+- All `$DOCKER_HUB_USERNAME` references should your Docker Hub Username
```
Comment (KimberlyFields) [inline tutorials/build-code/ci-tutorial-container-signing.md:146]:
```suggestion
- All `$DOCKER_HUB_USERNAME` references should your Docker Hub Username.
```

File: tutorials/build-code/ci-tutorial-container-signing.md [hunk 22]
```diff
@@ -0,0 +1,215 @@
...
+:::
+
+Click __Apply Changes__ to save the step and click __Save__ to save the pipeline.
```
Comment (KimberlyFields) [inline tutorials/build-code/ci-tutorial-container-signing.md:149]:
```suggestion
5. Select __Apply Changes__ to save the step, and then select __Save__ to save the pipeline.
```

File: tutorials/build-code/ci-tutorial-container-signing.md [hunk 23]
```diff
@@ -0,0 +1,215 @@
...
+
+![Final Pipeline](static/ci-tutorial-go-containers/go_pipeline_final.png)
+
+With those changes saved, you are ready to lint, test, build and push your __go__ application to container registry(DockerHub).
+
+## Run CI Pipeline
+
+As did earlier click __Run__ from the pipeline editor window,
```
Comment (KimberlyFields) [inline tutorials/build-code/ci-tutorial-container-signing.md:157]:
```suggestion
1. As you did earlier, select __Run__ from the pipeline editor window.
```

File: tutorials/build-code/ci-tutorial-container-signing.md [hunk 24]
```diff
@@ -0,0 +1,215 @@
...
+
+![Run Pipeline](static/ci-tutorial-go-containers/run_pipeline.png)
+
+Leaving everything to defaults namely __Git Branch__ and __Branch Name__ to be _main_, click __Run Pipeline__ to start the pipeline run.
```
Comment (KimberlyFields) [inline tutorials/build-code/ci-tutorial-container-signing.md:161]:
```suggestion
2. Leaving all the default entries in place, namely __Git Branch__ and __Branch Name__ to be _main_, select __Run Pipeline__ to start the pipeline run.
```

File: tutorials/build-code/ci-tutorial-container-signing.md [hunk 25]
```diff
@@ -0,0 +1,215 @@
...
+
+Once a successful run, head back to Docker Hub to check the image references!
+
+![Success](static/ci-tutorial-container-signing/go_pipeline_success.png)
+
+As you see the logs the container was built and the image reference is signed using our `$COSIGN_PRIVATE_KEY`.
+
+![Sign logs](static/ci-tutorial-container-signing/go_pipeline_sign_logs.png)
+
+## Verify Signed Image
```
Comment (KimberlyFields) [inline tutorials/build-code/ci-tutorial-container-signing.md:173]:
```suggestion
## Verify the signed image
```

File: tutorials/build-code/ci-tutorial-container-signing.md [hunk 26]
```diff
@@ -0,0 +1,215 @@
...
+
+To verify the image signature we use the public key `$COSIGN_PUBLIC_KEY` against the image `sha256` digest.
```
Comment (KimberlyFields) [inline tutorials/build-code/ci-tutorial-container-signing.md:175]:
```suggestion
To verify the image signature, use the public key `$COSIGN_PUBLIC_KEY` against the image `sha256` digest.
```

File: tutorials/build-code/ci-tutorial-container-signing.md [hunk 27]
```diff
@@ -0,0 +1,215 @@
...
+```
+
+Run the following command to verify the signature using our cosign public key,
+
+```shell
+cosign verify --key "$TUTORIAL_HOME/cosign.pub" \
+  "docker.io/$DOCKER_USERNAME/fruits-api@$IMAGE_REF"
+```
+
+The command should show an output as shown if the signature is valid, with `$DOCKER_USERNAME` and `$IMAGE_REF` replaced your docker hub username and image digest from previous step.
```
Comment (KimberlyFields) [inline tutorials/build-code/ci-tutorial-container-signing.md:190]:
```suggestion
The command should show an output as shown if the signature is valid, with `$DOCKER_USERNAME` and `$IMAGE_REF` replacing your Docker Hub username and image digest from the previous step.
```

File: tutorials/build-code/ci-tutorial-container-signing.md [hunk 28]
```diff
@@ -0,0 +1,215 @@
...
+```text
+Verification for index.docker.io/$DOCKER_USERNAME/fruits-api@$IMAGE_REF --
+The following checks were performed on each of these signatures:
+  - The cosign claims were validated
+  - The signatures were verified against the specified public key
+
+[{"critical":{"identity":{"docker-reference":"index.docker.io/$DOCKER_USERNAME/fruits-api"},"image":{"docker-manifest-digest":"$IMAGE_REF"},"type":"cosign container image signature"},"optional":null}]
+```
+
+In an ideal scenario the verification will be done by governance tools like [OPA](https://www.openpolicyagent.org/docs/latest/) or [kyverno](https://kyverno.io/). The integration with those tools can be done using Harness Platform via Continuous Deployment(CD) modules.
```
Comment (KimberlyFields) [inline tutorials/build-code/ci-tutorial-container-signing.md:201]:
```suggestion
In an ideal scenario the verification is done by governance tools like [OPA](https://www.openpolicyagent.org/docs/latest/) or [kyverno](https://kyverno.io/). The integration with those tools can be done using Harness Platform with Harness Continuous Delivery.
```

File: tutorials/build-code/ci-tutorial-container-signing.md [hunk 29]
```diff
@@ -0,0 +1,215 @@
...
+
+:::note Exercise
+
+Can you update the pipeline to a sign verification step ?
```
Comment (KimberlyFields) [inline tutorials/build-code/ci-tutorial-container-signing.md:205]:
```suggestion
Can you update the pipeline to a sign verification step?
```

File: tutorials/build-code/ci-tutorial-container-signing.md [hunk 30]
```diff
@@ -0,0 +1,215 @@
...
+
+__Clue__: Use `$IMAGE_REF` from __build and push__ step output variables.
+
+:::
+
+## Continuing on Your Continuous Integration Journey
```
Comment (KimberlyFields) [inline tutorials/build-code/ci-tutorial-container-signing.md:211]:
```suggestion
## Continuing on your continuous integration journey
```

File: tutorials/build-code/ci-tutorial-container-signing.md [hunk 31]
```diff
@@ -0,0 +1,215 @@
...
+
+You can now execute your builds whenever you want in a consistent fashion. Can modify the trigger to watch for SCM events so upon commit, for example, the Pipeline gets kicked off automatically. All of the objects you create are available for you to re-use. Lastly, you can even save your backing work / have it as part of your source code. Everything that you do in Harness is represented by YAML; feel free to store it as part of your project.
```
Comment (KimberlyFields) [inline tutorials/build-code/ci-tutorial-container-signing.md:213]:
```suggestion
You can now execute your builds whenever you want in a consistent fashion. You can modify the trigger to watch for SCM events so upon commit, for example, the pipeline is kicked off automatically. All of the objects you create are available for you to reuse. Lastly, you can even save your backing work / have it as part of your source code. Everything that you do in Harness is represented by YAML; feel free to store it as part of your project.
```

File: tutorials/build-code/ci-tutorial-container-signing.md [hunk 32]
```diff
@@ -0,0 +1,215 @@
...
+
+After you have built your artifact, the next step is to deploy your artifact. This is where Continuous Delivery steps in and make sure to check out some other [CD Tutorials](/tutorials/deploy-services#all-tutorials).
```
Comment (KimberlyFields) [inline tutorials/build-code/ci-tutorial-container-signing.md:215]:
```suggestion
After you have built your artifact, the next step is to deploy your artifact. This is where Continuous Delivery steps in. Make sure to check out some other [CD Tutorials](/tutorials/deploy-services#all-tutorials).
```

**Review by KimberlyFields — CHANGES_REQUESTED** [review]
Please use the "Add suggestion to batch" option to accept the suggestions so that all of the updates are made in one commit. If you use the "Commit suggestion" option on each suggestion, a new build kicks off after every commit and slows down the builds.

**Review by ravilach — APPROVED** [review]

**Pull request merged**